allensdk/api/cloud_cache/cloud_cache.py

```diff
@@ -0,0 +1,525 @@
+
+
+class LocalFileDescription(TypedDict):
+    local_path: pathlib.Path
+    exists: bool
+    file_attributes: CacheFileAttributes
+
+
+class CloudCache(object):
+    """
+    A class to handle the downloading and accessing of data served from a cloud
+    storage system
+
+    Parameters
+    ----------
+    cache_dir: str or pathlib.Path
+        Path to the directory where data will be stored on the local system
+
+    ***** THIS IS JUST A BASE CLASS AND CANNOT BE INSTANTIATED *****
+
+    Actual implementations of this class must implement
+    ===================================================
+
+    def _list_all_manifests(self) -> list:
+        Return a list of all of the file names of the manifests associated
+        with this dataset
+
+    def _download_manifest(self,
+                           manifest_name: str,
+                           output_stream: io.BytesIO):
+        Download a manifest from the dataset into output_stream.
+        Reset output_stream to the beginning
+
+        Parameters
+        ----------
+        manifest_name: str
+            The name of the manifest to load. Must be an element in
+            self.manifest_file_names
+
+        output_stream: io.BytesIO
+            A byte stream into which to load the manifest
+
+    def _download_file(self, file_attributes: CacheFileAttributes) -> bool:
+
+        Check if a file exists and is in the expected state.
+
+        If it is, return True.
+
+        If it is not, download the file, creating the directory
+        where the file is to be stored if necessary.
+
+        If the download is successful, return True.
+
+        If the download fails (file hash does not match expectation),
+        return False.
+
+        Parameters
+        ----------
+        file_attributes: CacheFileAttributes
+            Describes the file to download
+
+        Returns
+        -------
+        None
+
+    """
+
+    _bucket_name = None
+
+    def __init__(self, cache_dir):
+        self._manifest = Manifest(cache_dir)
+        self._manifest_file_names = self._list_all_manifests()
+
+    @property
+    def file_id_column(self) -> str:
+        """
+        The column in the metadata files used to uniquely
+        identify data files
+        """
+        return self._manifest.file_id_column
+
+    @property
+    def version(self) -> str:
+        """
+        The version of the dataset currently loaded
+        """
+        return self._manifest.version
+
+    @property
+    def metadata_file_names(self) -> list:
+        """
+        List of metadata file names associated with this dataset
+        """
+        return self._manifest.metadata_file_names
+
+    @property
+    def manifest_file_names(self) -> list:
+        """
+        Sorted list of manifest file names associated with this
+        dataset
+        """
+        return copy.deepcopy(self._manifest_file_names)
+
+    def load_manifest(self, manifest_name: str):
```

Comment:
and then, these would have the @abstractmethod decorator on them and/or raise NotImplementedError
https://docs.python.org/3/library/abc.html#abc.abstractmethod